src/dtree.ml

```diff
@@ -0,0 +1,515 @@
+    { values = List.map (fun r ->
+      let term_pos = Position.mark r.Terms.lhs in
+      { lhs = term_pos ; rhs = ref r.Terms.rhs }) rs
+    ; var_catalogue = [] }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> m.values = []
+
+  (** [get_col n m] retrieves column [n] of matrix [m].  There is some
+      processing because all rows do not have the same length. *)
+  let get_col : int -> t -> line = fun ind { values ; _ } ->
+    let opcol = List.fold_left (fun acc { lhs = li ; _ } ->
+        List.nth_opt li ind :: acc) []
+        values in
+    let rem = List.filter (function None -> false | Some(_) -> true) opcol in
+    List.map (function Some(e) -> e | None -> assert false) rem
+
+  (** [select m i] keeps the columns of [m] whose index are in [i]. *)
+  let select : t -> int array -> t = fun m indexes ->
+    { m with
+      values = List.map (fun rul ->
+        { rul with
+          lhs = Array.fold_left (fun acc i -> List.nth rul.lhs i :: acc)
+            [] indexes }) m.values }
+
+  (** [swap p i] swaps the first column with the [i]th one. *)
+  let swap : t -> int -> t = fun pm c ->
+    { pm with values = List.map (fun rul ->
+      { rul with lhs = List.bring c rul.lhs }) pm.values }
+
+  (** [cmp c d] compares columns [c] and [d] returning: +1 if [c > d], 0 if
+      [c = d] or -1 if [c < d]; where [<], [=] and [>] are defined according
+      to a heuristic.  *)
+  let cmp : line -> line -> int = fun _ _ -> 0
+
+  (** [pick_best m] returns the index of the best column of matrix [m]
+      according to a heuristic. *)
+  let pick_best : t -> int = fun _ -> 0
+
+  (** [is_pattern t] returns whether a term [t] is considered as a pattern *)
+  let is_pattern : term -> bool = function
+    | Patt(_, _, _) -> false
+    | _             -> true
+
+  (** [exhausted r] returns whether rule [r] can be further pattern matched or
+      if it is ready to yield the action.  A rule is exhausted when its left
+      hand side is composed only of wildcards. *)
+  let exhausted : rule -> bool = fun { lhs ; _ }->
+    let rec loop = function
+      | []                            -> true
+      | (x, _) :: _ when is_pattern x -> false
+      | _ :: xs                       -> loop xs in
+    loop lhs
+
+  (** [can_switch_on p k] returns whether a switch can be carried out on
+      column [k] in matrix [p] *)
+  let can_switch_on : t -> int -> bool = fun { values ; _ } k ->
+    let rec loop : rule list -> bool = function
+      | []      -> true
+      | r :: rs ->
+        begin
+          match List.nth_opt r.lhs k with
+          | None       -> loop rs
+          | Some(t, _) -> is_pattern t
+        end in
+    loop values
+
+  (** [discard_patt_free m] returns the list of indexes of columns containing
+      terms that can be matched against (discard pattern-free columns). *)
+  let discard_patt_free : t -> int array = fun m ->
+    let ncols = List.fold_left (fun acc { lhs ; _ } ->
```

Comment:
Use a let for long anonymous functions. It's more readable.

Reply:
solved in c3a35d9 using `extremum`